docs/engineering/developer-guide.md

```diff
@@ -12,7 +12,123 @@ To build and test this repo, run `build.ps1` on Windows, run `build.sh` on Mac O
 
 ## Coding Style
 
-We use [C# Coding Style](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md) recommended by dotnet team. Stylecop and FxCop have been enabled for this project to enforce some of the rules.
+### C#
+We follow [C# Coding Style](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md) recommended by dotnet team. Stylecop and FxCop have been enabled for this project to enforce some of the rules.
 
+Besides that, we have some recommended but not mandatory (that is, not enabled by StyleCop/FxCop) rules as illustrated below.
 
+#### Performance Consideration
+* **DO** use `sealed` for private classes if they are not to be inherited.
+* **DO** add `readonly` to fields if they do not tend to be changed.
+* **DO** use `static` methods if it is not instance relevant.
+* **DO** use `RegexOptions.Compiled` for **readonly** *Regex*.
+
+#### Unit tests and functional tests
+##### Assembly naming
+The unit tests for the `Microsoft.Foo` assembly live in the `Microsoft.Foo.Tests` assembly.
+
+The functional tests for the `Microsoft.Foo` assembly live in the `Microsoft.Foo.FunctionalTests` assembly.
+
+In general there should be exactly one unit tests assembly for each product runtime assembly. In general there should be one functional tests assembly per repo. Exceptions can be made for both.
+
+##### Unit test class naming
+Test class names end with `Test` suffix and live in the same namespace as the class being tested. For example, the unit tests for the `Microsoft.Foo.Boo` class would be in a `Microsoft.Foo.BooTest` class in the unit tests assembly `Microsoft.Foo.Tests`.
+
+##### Unit test method naming
+Unit test method names must be descriptive about *what developers are testing, under what conditions, and what the expectations are*. Pascal casing and underscores can be used to improve readability. The following test names are correct:
+
+```cs
+PublicApiArgumentsShouldHaveNotNullAnnotation
+Public_api_arguments_should_have_not_null_annotation
```

Comment:
The current StyleCop rule does not support this for test. src/test now shares the same rule

Reply:
is it possible to define another set of rules for test projects?

---
In reply to: [180969320](https://github.com/dotnet/docfx/pull/2629#discussion_r180969320) [](ancestors = 180969320)